File: 0116-填充每个节点的下一个右侧节点指针/Solution2.java

class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {
    }

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};

public class Solution {
    public Node connect(Node root) {
        helper(root);
        return root;
    }
    private void helper(Node root){
        if(root == null || root.left == null){ return; }
        helper(root.left);
        helper(root.right);
        Node p=root.left, q=root.right;
        while(p != null){
            p.next = q;
            p = p.right;
            q = q.left;
        }
    }
}

File: 0110-平衡二叉树/solution2.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.cache = {}

    def isBalanced(self, root: TreeNode) -> bool:
        if not root:
            return True
        left = self.height(root.left)
        right = self.height(root.right)
        if abs(left - right) > 1:
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)

    def height(self, root):
        if root is None:
            return 0
        if root in self.cache:
            return self.cache[root]
        left = self.height(root.left)
        right = self.height(root.right)
        ret = 1 + max(left, right)
        self.cache[root] = ret
        return ret